README.md
# FasterImage [![CircleCI](https://img.shields.io/circleci/project/github/willwashburn/fasterimage.svg?style=flat-square)](https://circleci.com/gh/willwashburn/fasterimage) [![Coveralls](https://img.shields.io/coveralls/willwashburn/fasterimage.svg?maxAge=259200&style=flat-square)](https://coveralls.io/github/willwashburn/fasterimage) [![Minimum PHP Version](https://img.shields.io/badge/php-%3E%3D%205.4-8892BF.svg?style=flat-square)](https://php.net/) [![Packagist Version](https://img.shields.io/packagist/v/fasterimage/fasterimage.svg?style=flat-square)](https://packagist.org/packages/fasterimage/fasterimage) [![Packagist Downloads](https://img.shields.io/packagist/dt/fasterimage/fasterimage.svg?style=flat-square)](https://packagist.org/packages/fasterimage/fasterimage/stats) [![License](https://img.shields.io/packagist/l/fasterimage/fasterimage.svg?style=flat-square)](https://github.com/willwashburn/fasterimage/LICENSE)

FasterImage finds the dimensions or filetype of a remote image file given its uri by fetching as little as needed, based on the excellent [Ruby implementation by Stephen Sykes](https://github.com/sdsykes/fastimage) and [PHP implementation by Tom Moor](https://github.com/tommoor/fastimage).

FasterImage uses the curl_muli* suite to run requests in parallel. Currently supports JPG, GIF, PNG, WEBP, BMP, PSD, TIFF, SVG, and ICO files.

## Usage
```php

        $client = new \FasterImage\FasterImage();
        
        $images = $client->batch([
            'http://wwww.example.com/image1.jpg',
            'http://wwww.example.com/image2.gif',
            'http://wwww.example.com/image3.png',
            'http://wwww.example.com/image4.bmp',
            'http://wwww.example.com/image5.tiff',
            'http://wwww.example.com/image6.psd',
            'http://wwww.example.com/image7.webp',
            'http://wwww.example.com/image8.ico',
            'http://wwww.example.com/image9.cur',
            'http://wwww.example.com/image10.svg'
        ]);
        
        foreach ($images as $image) {
            list($width,$height) = $image['size'];
        }
```

## Install

```composer require fasterimage/fasterimage```

Alternatively, add ```"fasterimage/fasterimage": "~1.5"``` to your composer.json

## Changelog

* v1.5.0 - Fallback support when curl_multi_init() is not available
* v1.4.0 - Add support for parsing dimensions from SVG images
* v1.3.0 - Add ability for user agent, buffer size, and SSL host/peer verification to be overridden
* v1.2.1 - Limit isRotated to only check for valid orientation values
* v1.2.0 - Add option to include content-length in result set
* v1.1.2 - Update Accept header to accept images
* v1.1.1 - Properly handle jpeg's with corrupted Exif tags 
* v1.1.0 - Return message in return array when curl fails
* v1.0.3 - Use external stream package
* v1.0.2 - Fail invalid image exceptions gracefully when using batch requests
* v1.0.1 - Support PHP v5.4+
* v1.0.0 - **Stable Release** - Support for `.PSD`, `.ICO` + `.CUR`
* v0.0.7 - Remove support for PHP v5.4
* v0.0.6 - Add option to set timeout of requests, support for EXIF in .jpgs, better support for .bmp (including negative height bitmaps) and normalized response indexes for all file types
* v0.0.5 - Support for `.webp`
* v0.0.4 - Support for `.tiff` and exceptions for unknown file types
* v0.0.3 - Force curl to follow redirects so you get less bad responses
* v0.0.2 - Update curl headers to mimic browser so you get less bad responses
* v0.0.1 - Support for `.jpg`, `.bmp`, `.gif`, `.png` and parallel requests

## References

* https://github.com/sdsykes/fastimage
* https://github.com/tommoor/fastimage
* http://pennysmalls.com/find-jpeg-dimensions-fast-in-pure-ruby-no-ima
* http://snippets.dzone.com/posts/show/805
* http://www.anttikupila.com/flash/getting-jpg-dimensions-with-as3-without-loading-the-entire-file/
* http://imagesize.rubyforge.org/


CHANGELOG.md
## Changelog
### 1.5.0 (2019-05-25)
* Fallback support when curl_multi_init() is not available

### 1.4.0 (2019-03-07)
* Add support for parsing dimensions from SVG images

### 1.3.0 (2019-03-06)
* Add ability for user agent, buffer size, and SSL host/peer verification to be overridden

### 1.2.1 (2019-02-18)
* Limit isRotated to only check for valid orientation values

### 1.2.0 (2018-03-07)
* Add option to include content-length in result set

### 1.1.2 (2018-03-07)
* Update Accept header to accept images

### 1.1.1 (2016-11-04)
* Properly handle jpeg's with corrupted Exif tags 

### 1.1.0 (2016-04-30)
* Return message in return array when curl fails

### 1.0.3 (2016-04-9)
* Use external stream package

### 1.0.2 (2015-08-27)
* Fail invalid image exceptions gracefully when using batch requests

### 1.0.1 (2015-05-16)

* Support PHP v5.4+

### 1.0.0 (2015-05-16) [stable release!]

* Support for .PSD
* Support for .ICO + .CUR

### 0.0.7 (2015-05-16)

* Only support PHP v5.5+ for now :/

### 0.0.6 (2015-05-11)

* Add option to set timeout of requests
* Support for EXIF in .jpgs
* Better support for .bmp (including negative height bitmaps)
* Normalized response indexes for all file types

### 0.0.5 (2015-05-09)

* Support for .webp

### 0.0.4 (2015-05-07)

* Support for .tiff
* Start throwing exceptions for unknown file types

### 0.0.3 (2015-05-07)

* Force curl to follow redirects so you get less bad responses

### 0.0.2 (2015-05-06)

* Update curl headers to mimic browser so you get less bad responses

### 0.0.1 (2015-05-06) [alpha release!]

* Support for .jpg
* Support for .bmp
* Support for .gif
* Support for .png
* Support for parallel requests


tests/FasterImageTest.php
<?php
chdir(__DIR__);
include('../vendor/autoload.php');

/**
 * Class FasterImageTest
 */
class FasterImageTest extends \PHPUnit\Framework\TestCase
{
    use \phpmock\phpunit\PHPMock;

    /**
     * Set up.
     */
    protected function setUp() {
        parent::setUp();

        // Mock function_exists() to return false for all curl_multi_* functions when PHPUnit is invoked with DISABLE_CURL_MULTI=1 environment variable.
        if ( isset( $_ENV['DISABLE_CURL_MULTI'] ) && true === filter_var( $_ENV['DISABLE_CURL_MULTI'], FILTER_VALIDATE_BOOLEAN ) ) {
            $function_exists = $this->getFunctionMock('FasterImage', 'function_exists');
            $function_exists->expects($this->atLeastOnce())->willReturnCallback(
                function($function) {
                    if ( is_string($function) && 0 === strpos( $function, 'curl_multi_' ) ) {
                        return false;
                    } else {
                        return function_exists($function);
                    }
                }
            );
        }
    }

    /**
     * @throws Exception
     */
    public function test_invalid_images_return_failed()
    {
        $uris = [
            'http://example.com/foobarimage.jpg',
            'https://example.com/foobarimage.jpg',
            'sdfsdfdsfds',
        ];

        $client = new \FasterImage\FasterImage();
        $images = $client->batch($uris);

        foreach ( $images as $uri => $image ) {
            $this->assertArrayHasKey('size', $image);
            $this->assertEquals('failed', $image['size']);
        }
    }

    /**
     * @throws \Exception
     */
    public function test_batch_returns_size_and_type()
    {
        $data = $this->linksProvider();

        $expected = [];
        $uris     = [];

        foreach ( $data as $link ) {
            $uri    = current($link);
            $width  = next($link);
            $height = next($link);
            $type   = next($link);

            $uris[]         = $uri;
            $expected[$uri] = compact('width', 'height', 'type');
        }

        $client = new \FasterImage\FasterImage();
        $client->setTimeout(120);
        $images = $client->batch($uris);

        foreach ( $images as $uri => $image ) {
            $this->assertArrayHasKey('type', $image, "$uri is missing type: " . print_r($image, true));
            $this->assertEquals($expected[$uri]['type'], $image['type'], "Failed to get the right type for $uri");
            $this->assertArrayHasKey('size', $image, "There is no size defined for $uri " . print_r($image, true));
            $this->assertEquals($expected[$uri]['width'], $image['size'][0], "Failed to get the right width for $uri " . print_r($image, true));
            $this->assertEquals($expected[$uri]['height'], $image['size'][1], "Failed to get the right height for $uri " . print_r($image, true));
        }
    }

    /**
     * @return array
     */
    public function linksProvider()
    {
        return array(
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/exif_orientation.jpg', 600, 450, 'jpeg'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/favicon.ico', 16, 16, 'ico'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/infinite.jpg', 160, 240, 'jpeg'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/man.ico', 48, 48, 'ico'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/orient_2.jpg', 230, 408, 'jpeg'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test.bmp', 40, 27, 'bmp'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test.cur', 32, 32, 'cur'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test.gif', 17, 32, 'gif'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test.jpg', 882, 470, 'jpeg'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test.png', 30, 20, 'png'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test.psd', 17, 32, 'psd'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test.tiff', 85, 67, 'tiff'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test2.bmp', 1920, 1080, 'bmp'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test2.jpg', 250, 188, 'jpeg'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test2.tiff', 333, 225, 'tiff'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test3.jpg', 630, 367, 'jpeg'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/test4.jpg', 1485, 1299, 'jpeg'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/webp_vp8.webp', 550, 368, 'webp'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/webp_vp8l.webp', 386, 395, 'webp'],
            ['https://github.com/sdsykes/fastimage/raw/master/test/fixtures/webp_vp8x.webp', 386, 395, 'webp'],
            ['http://ketosizeme.com/wp-content/uploads/2016/11/Keto-Corn-Dog-Recipe-Low-Carb-High-Fat-.jpg', 700, 467, 'jpeg'],
            ['http://gluesticksgumdrops.com/wp-content/uploads/2015/03/how-to-find-more-time-to-read-to-your-kids.jpg', 700, 1000, 'jpeg'],
            ['https://s.w.org/images/core/emoji/11/svg/1f642.svg', 36, 36, 'svg'], // With viewBox only.
            ['https://gist.github.com/westonruter/0d66e6629526fc820a31fd43cf325376/raw/ed26b6d60b3d1c07a82907553795e6535c12da89/smiley-oval-vertical-padding.svg', 36, 96, 'svg'],
            ['https://gist.github.com/westonruter/0d66e6629526fc820a31fd43cf325376/raw/4c3b8d7fdad3233d669c05c7313c0503c8a3c8ed/smiley-with-dimensions-and-offcenter-viewbox.svg', 36, 42, 'svg'],
            ['https://gist.github.com/westonruter/0d66e6629526fc820a31fd43cf325376/raw/23651d8fec1a22707632f687b742700896daf54f/svg-tag-commented-out.svg', 38, 44, 'svg'],
        );
    }
}


examples/index.php
<?php

$time = microtime(true);

$images = [
    'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQQC3-MnPUUr3Z-pNsIl7Z33BXTUG0DtEzmbtjXV_hNhBnup5QyYPeUKpo',
    'http://cdn.shopify.com/s/files/1/0224/1915/files/bunny.jpg?22110',
    'http://36.media.tumblr.com/a5dbdd0882a3de34b48f9109599a3060/tumblr_nnp84siJ5x1qjcdw9o1_1280.jpg',
];


$time = microtime(true);
$fast  = new FasterImage\FasterImage();
$sizes = $fast->batch($images);

foreach ( $sizes as $image ) {

    list($width, $height) = $image['size'];
    echo "FasterImage: \n";
    echo "Width: " . $width . "px Height: " . $height . "px in " . (microtime(true) - $time) . " seconds \n";

}

echo "FasterImage for all three: " . $fasterimage = (microtime(true) - $time) . " seconds \n";

$time = microtime(true);
foreach ( $images as $image ) {

    $time = microtime(true);
    list($width, $height) = getimagesize($image);
    echo "getimagesize: \n";
    echo "Width: " . $width . "px Height: " . $height . "px in " . (microtime(true) - $time) . " seconds \n";
}

echo "getimagesize for all three: " . $getimagesize = (microtime(true) - $time) . " seconds \n";
echo PHP_EOL;
echo $getimagesize - $fasterimage ." seconds faster";
exit;

src/FasterImage/Stream.php
<?php namespace FasterImage;

use FasterImage\Exception\StreamBufferTooSmallException;

/**
 * Class Stream
 *
 * @package FasterImage
 */
class Stream implements StreamableInterface
{
    /**
     * The string that we have downloaded so far
     */
    protected $stream_string = '';

    /**
     * The pointer in the string
     *
     * @var int
     */
    protected $strpos = 0;

    /**
     * Get characters from the string but don't move the pointer
     *
     * @param $characters
     *
     * @return string
     * @throws \FasterImage\Exception\StreamBufferTooSmallException
     */
    public function peek($characters)
    {
        if ( strlen($this->stream_string) < $this->strpos + $characters ) {
            throw new StreamBufferTooSmallException('Not enough of the stream available.');
        }

        return substr($this->stream_string, $this->strpos, $characters);
    }

    /**
     * Get Characters from the string
     *
     * @param $characters
     *
     * @return string
     * @throws \FasterImage\Exception\StreamBufferTooSmallException
     */
    public function read($characters)
    {
        $result = $this->peek($characters);

        $this->strpos += $characters;

        return $result;
    }

    /**
     * Resets the pointer to the 0 position
     *
     * @return mixed
     */
    public function resetPointer()
    {
        $this->strpos = 0;
    }

    /**
     * Append to the stream string
     *
     * @param $string
     */
    public function write($string)
    {
        $this->stream_string .= $string;
    }
}

src/FasterImage/ExifParser.php
<?php namespace FasterImage;

use FasterImage\Exception\InvalidImageException;

/**
 * Class ExifParser
 *
 * @package FasterImage
 */
class ExifParser
{
    /**
     * @var int
     */
    protected $width;
    /**
     * @var  int
     */
    protected $height;

    /**
     * @var
     */
    protected $short;

    /**
     * @var
     */
    protected $long;

    /**
     * @var  StreamableInterface
     */
    protected $stream;

    /**
     * @var int
     */
    protected $orientation;

    /**
     * @param \FasterImage\StreamableInterface $stream
     */
    public function __construct(StreamableInterface $stream)
    {
        $this->stream = $stream;
        $this->parseExifIfd();
    }

    /**
     * @return int
     */
    public function getHeight()
    {
        return $this->height;
    }

    /**
     * @return int
     */
    public function getWidth()
    {
        return $this->width;
    }

    /**
     * @return bool
     */
    public function isRotated()
    {
        return (! empty($this->orientation) && $this->orientation >= 5);
    }

    /**
     * @return bool
     * @throws \FasterImage\Exception\InvalidImageException
     */
    protected function parseExifIfd()
    {
        $byte_order = $this->stream->read(2);

        switch ( $byte_order ) {
            case 'II':
                $this->short = 'v';
                $this->long  = 'V';
                break;
            case 'MM':
                $this->short = 'n';
                $this->long  = 'N';
                break;
            default:
                throw new InvalidImageException;
                break;
        }

        $this->stream->read(2);

        $offset = current(unpack($this->long, $this->stream->read(4)));

        $this->stream->read($offset - 8);

        $tag_count = current(unpack($this->short, $this->stream->read(2)));

        for ( $i = $tag_count; $i > 0; $i-- ) {

            $type = current(unpack($this->short, $this->stream->read(2)));
            $this->stream->read(6);
            $data = current(unpack($this->short, $this->stream->read(2)));

            switch ( $type ) {
                case 0x0100:
                    $this->width = $data;
                    break;
                case 0x0101:
                    $this->height = $data;
                    break;
                case 0x0112:
                    $this->orientation = $data;
                    break;
            }

            if ( isset($this->width) && isset($this->height) && isset($this->orientation) ) {
                return true;
            }

            $this->stream->read(2);
        }

        return false;
    }
}

src/FasterImage/ImageParser.php
<?php namespace FasterImage;

use WillWashburn\Stream\Exception\StreamBufferTooSmallException;
use WillWashburn\Stream\Stream;
use WillWashburn\Stream\StreamableInterface;

/**
 * Parses the stream of the image and determines the size and type of the image
 *
 * @package FasterImage
 */
class ImageParser
{
    /**
     * The type of image we've determined this is
     *
     * @var string
     */
    protected $type;
    /**
     * @var StreamableInterface $stream
     */
    private $stream;

    /**
     * ImageParser constructor.
     *
     * @param StreamableInterface $stream
     */
    public function __construct(StreamableInterface & $stream)
    {
        $this->stream = $stream;
    }

    /**
     * @return array|bool|null
     */
    public function parseSize()
    {
        $this->stream->resetPointer();

        switch ( $this->type ) {
            case 'png':
                return $this->parseSizeForPNG();
            case 'ico':
            case 'cur':
                return $this->parseSizeForIco();
            case 'gif':
                return $this->parseSizeForGIF();
            case 'bmp':
                return $this->parseSizeForBMP();
            case 'jpeg':
                return $this->parseSizeForJPEG();
            case 'tiff':
                return $this->parseSizeForTiff();
            case 'psd':
                return $this->parseSizeForPSD();
            case 'webp':
                return $this->parseSizeForWebp();
            case 'svg':
                return $this->parseSizeForSvg();
        }

        return null;
    }

    /**
     * @return array
     */
    public function parseSizeForIco()
    {
        $this->stream->read(6);

        $b1 = $this->getByte();
        $b2 = $this->getByte();

        return [
            $b1 == 0 ? 256 : $b1,
            $b2 == 0 ? 256 : $b2
        ];
    }

    /**
     * @return array
     */
    protected function parseSizeForPSD() {

        $this->stream->read(14);
        $sizes = unpack("N*",$this->stream->read(12));

        return [
            $sizes[2],
            $sizes[1]
        ];
    }

    /**
     * Reads and returns the type of the image
     *
     * @return bool|string
     */
    public function parseType()
    {
        if ( ! $this->type ) {
            $this->stream->resetPointer();

            switch ( $this->stream->read(2) ) {
                case "BM":
                    return $this->type = 'bmp';
                case "GI":
                    return $this->type = 'gif';
                case chr(0xFF) . chr(0xd8):
                    return $this->type = 'jpeg';
                case "\0\0":
                    switch ( $this->readByte($this->stream->peek(1)) ) {
                        case 1:
                            return $this->type = 'ico';
                        case 2:
                            return $this->type = 'cur';
                    }

                    return false;

                case chr(0x89) . 'P':
                    return $this->type = 'png';
                case "RI":
                    if ( substr($this->stream->read(10), 6, 4) == 'WEBP' ) {
                        return $this->type = 'webp';
                    }

                    return false;
                case'8B':
                    return $this->type = 'psd';
                case "II":
                case "MM":
                    return $this->type = 'tiff';
                default:
                    $this->stream->resetPointer();

                    // Keep reading bytes until we find '<svg'.
                    while ( true ) {
                        $byte = $this->stream->read( 1 );
                        if ( '<' === $byte && 'svg' === $this->stream->peek( 3 ) ) {
                            $this->type = 'svg';
                            return $this->type;
                        }
                    }
                    return false;
            }
        }

        return $this->type;
    }

    /**
     * @return array
     */
    protected function parseSizeForBMP()
    {
        $chars = $this->stream->read(29);
        $chars = substr($chars, 14, 14);
        $type  = unpack('C', $chars);

        $size = (reset($type) == 40) ? unpack('l*', substr($chars, 4)) : unpack('l*', substr($chars, 4, 8));

        return [
            current($size),
            abs(next($size))
        ];
    }

    /**
     * @return array
     */
    protected function parseSizeForGIF()
    {
        $chars = $this->stream->read(11);

        $size = unpack("S*", substr($chars, 6, 4));

        return [
            current($size),
            next($size)
        ];
    }

    /**
     * @return array|bool
     */
    protected function parseSizeForJPEG()
    {
        $state = null;

        while ( true ) {
            switch ( $state ) {
                default:
                    $this->stream->read(2);
                    $state = 'started';
                    break;

                case 'started':
                    $b = $this->getByte();
                    if ( $b === false ) return false;

                    $state = $b == 0xFF ? 'sof' : 'started';
                    break;

                case 'sof':
                    $b = $this->getByte();

                    if ( $b === 0xe1 ) {
                        $data = $this->stream->read($this->readInt($this->stream->read(2)) - 2);

                        $stream = new Stream;
                        $stream->write($data);

                        if ( $stream->read(4) === 'Exif' ) {

                            $stream->read(2);

                            // Some Exif data is broken/wrong so we'll ignore
                            // any exceptions here
                            try {
                                $exif = new ExifParser($stream);
                            } catch (\Exception $e) {}

                        }

                        break;
                    }

                    if ( in_array($b, range(0xe0, 0xef)) ) {
                        $state = 'skipframe';
                        break;
                    }

                    if ( in_array($b, array_merge(range(0xC0, 0xC3), range(0xC5, 0xC7), range(0xC9, 0xCB), range(0xCD, 0xCF))) ) {
                        $state = 'readsize';
                        break;
                    }
                    if ( $b == 0xFF ) {
                        $state = 'sof';
                        break;
                    }

                    $state = 'skipframe';
                    break;

                case 'skipframe':
                    $skip = $this->readInt($this->stream->read(2)) - 2;
                    $this->stream->read($skip);
                    $state = 'started';
                    break;

                case 'readsize':
                    $c = $this->stream->read(7);

                    $size = array($this->readInt(substr($c, 5, 2)), $this->readInt(substr($c, 3, 2)));

                    if ( isset($exif) && $exif->isRotated() ) {
                        return array_reverse($size);
                    }

                    return $size;
            }
        }

        return false;
    }

    /**
     * @return array
     */
    protected function parseSizeForPNG()
    {
        $chars = $this->stream->read(25);

        $size = unpack("N*", substr($chars, 16, 8));

        return [
            current($size),
            next($size)
        ];

    }

    /**
     * @return array|bool
     * @throws \FasterImage\Exception\InvalidImageException
     * @throws StreamBufferTooSmallException
     */
    protected function parseSizeForTiff()
    {
        $exif = new ExifParser($this->stream);

        if ( $exif->isRotated() ) {
            return [$exif->getHeight(), $exif->getWidth()];
        }

        return [$exif->getWidth(), $exif->getHeight()];
    }

    /**
     * @return null
     * @throws StreamBufferTooSmallException
     */
    protected function parseSizeForWebp()
    {
        $vp8 = substr($this->stream->read(16), 12, 4);
        $len = unpack("V", $this->stream->read(4));

        switch ( trim($vp8) ) {

            case 'VP8':
                $this->stream->read(6);

                $width  = current(unpack("v", $this->stream->read(2)));
                $height = current(unpack("v", $this->stream->read(2)));

                return [
                    $width & 0x3fff,
                    $height & 0x3fff
                ];

            case 'VP8L':
                $this->stream->read(1);

                $b1 = $this->getByte();
                $b2 = $this->getByte();
                $b3 = $this->getByte();
                $b4 = $this->getByte();

                $width  = 1 + ((($b2 & 0x3f) << 8) | $b1);
                $height = 1 + ((($b4 & 0xf) << 10) | ($b3 << 2) | (($b2 & 0xc0) >> 6));

                return [$width, $height];

            case 'VP8X':

                $flags = current(unpack("C", $this->stream->read(4)));

                $b1 = $this->getByte();
                $b2 = $this->getByte();
                $b3 = $this->getByte();
                $b4 = $this->getByte();
                $b5 = $this->getByte();
                $b6 = $this->getByte();

                $width = 1 + $b1 + ($b2 << 8) + ($b3 << 16);

                $height = 1 + $b4 + ($b5 << 8) + ($b6 << 16);

                return [$width, $height];
            default:
                return null;
        }

    }

    /**
     * Parse size for SVG.
     *
     * @return array|null Size or null.
     */
    protected function parseSizeForSvg()
    {
        $this->stream->resetPointer();

        // Keep reading bytes until
        $inside = false;
        $markup = '';
        while ( true ) {
            $byte = $this->stream->read(1);

            // Open a tag if not in a tag and the byte is '<'.
            if ( ! $inside && '<' === $byte ) {
                $inside = true;
                $markup .= $byte;
            }
            // Close the current tag if the tag is open, the byte is '>', and the last characters weren't a comment token.
            elseif ( $inside && '>' === $byte && '--' !== substr( $markup, -2 ) ) {

                $inside = false;
                $markup .= $byte;

                // Break the loop if this tag started with '<svg', as we have now found the SVG start tag.
                if ( '<svg' === strtolower( substr( $markup, 0, 4 ) ) ) {
                    break;
                }

                // Clear out the markup since we consumed the end of the tag/comment.
                $markup = '';
            }
            // Append the bte to the current tag if the tag is open.
            elseif ( $inside ) {
                $markup .= $byte;
            }
        }

        $width = null;
        $height = null;
        if ( preg_match( '/\swidth=([\'"])(\d+(\.\d+)?)(px)?\1/', $markup, $matches ) ) {
            $width = floatval( $matches[2] );
        }
        if ( preg_match( '/\sheight=([\'"])(\d+(\.\d+)?)(px)?\1/', $markup, $matches ) ) {
            $height = floatval( $matches[2] );
        }
        if ( $width && $height ) {
            return [ $width, $height ];
        }
        if ( preg_match( '/\sviewBox=([\'"])[^\1]*(?:,|\s)+(?P<width>\d+(?:\.\d+)?)(?:px)?(?:,|\s)+(?P<height>\d+(?:\.\d+)?)(?:px)?\s*\1/', $markup, $matches ) ) {
            return [
                floatval( $matches['width'] ),
                floatval( $matches['height'] )
            ];
        }
        return null;
    }

    /**
     * @return mixed
     */
    private function getByte()
    {
        return $this->readByte($this->stream->read(1));
    }

    /**
     * @param $string
     *
     * @return mixed
     */
    private function readByte($string)
    {
        return current(unpack("C", $string));
    }

    /**
     * @param $str
     *
     * @return int
     */
    private function readInt($str)
    {
        $size = unpack("C*", $str);

        return ($size[1] << 8) + $size[2];
    }
}
